sdk/storage_blobs/src/blob_client.rs

```diff
@@ -41,19 +47,91 @@ impl BlobClient {
             container_name: container_name,
             blob_name: blob_name,
             url: Url::parse(&blob_url).expect("Something went wrong with URL parsing!"),
-            pipeline: BlobClient::build_pipeline(credential, options.client_options),
+            pipeline: BlobClient::<Unset>::build_pipeline(credential, options.client_options),
+            state: Unset,
+        }
+    }
+
+    // We can make this alot better by adding the Default implements
+    fn as_block_blob(self) -> BlobClient<Block> {
+        BlobClient {
+            account_name: self.account_name,
+            credential: self.credential,
+            container_name: self.container_name,
+            blob_name: self.blob_name,
+            url: self.url,
+            pipeline: self.pipeline,
+            state: Block,
+        }
+    }
+
+    fn as_append_blob(self) -> BlobClient<Append> {
+        BlobClient {
+            account_name: self.account_name,
+            credential: self.credential,
+            container_name: self.container_name,
+            blob_name: self.blob_name,
+            url: self.url,
+            pipeline: self.pipeline,
+            state: Append,
+        }
+    }
+
+    fn as_page_blob(self) -> BlobClient<Page> {
+        BlobClient {
+            account_name: self.account_name,
+            credential: self.credential,
+            container_name: self.container_name,
+            blob_name: self.blob_name,
+            url: self.url,
+            pipeline: self.pipeline,
+            state: Page,
         }
     }
 
     // This will handle appending container and blob name
     fn build_blob_url(base_url: &str, container_name: &str, blob_name: &str) -> String {
         base_url.to_owned() + container_name + "/" + blob_name
     }
+}
+
+impl<T: BlobKind> BlobClient<T> {
+    pub async fn upload_blob(&self, data: Bytes, blob_type: Option<String>) -> Result<Response> {
+        // Build the upload properties request itself
+        let mut request = Request::new(self.url.to_owned(), Method::Put); // This is technically cloning
+
+        // This will be better when we can use an Enum or something
+        match blob_type {
+            Some(b) => {
+                if ["BlockBlob", "PageBlob", "AppendBlob"].contains(&b.as_str()) {
+                    request.insert_header("x-ms-blob-type", b)
+                } else {
+                    request.insert_header("x-ms-blob-type", "BlockBlob")
+                }
+            }
+            None => request.insert_header("x-ms-blob-type", "BlockBlob"),
```

Comment:
This is not as cool as I thought it would be 😄 probably going to refactor this out -- too much Python brain trying to `keyword`-esque functionality here